doc/musig-reference.py

```diff
@@ -0,0 +1,284 @@
+
+# Points are tuples of X and Y coordinates and the point at infinity is
+# represented by the None keyword.
+G = (0x79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798, 0x483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8)
+
+Point = Tuple[int, int]
+
+# This implementation can be sped up by storing the midstate after hashing
+# tag_hash instead of rehashing it all the time.
+def tagged_hash(tag: str, msg: bytes) -> bytes:
+    tag_hash = hashlib.sha256(tag.encode()).digest()
+    return hashlib.sha256(tag_hash + tag_hash + msg).digest()
+
+def is_infinite(P: Optional[Point]) -> bool:
+    return P is None
+
+def x(P: Point) -> int:
+    assert not is_infinite(P)
+    return P[0]
+
+def y(P: Point) -> int:
+    assert not is_infinite(P)
+    return P[1]
+
+def point_add(P1: Optional[Point], P2: Optional[Point]) -> Optional[Point]:
+    if P1 is None:
+        return P2
+    if P2 is None:
+        return P1
+    if (x(P1) == x(P2)) and (y(P1) != y(P2)):
+        return None
+    if P1 == P2:
+        lam = (3 * x(P1) * x(P1) * pow(2 * y(P1), p - 2, p)) % p
+    else:
+        lam = ((y(P2) - y(P1)) * pow(x(P2) - x(P1), p - 2, p)) % p
+    x3 = (lam * lam - x(P1) - x(P2)) % p
+    return (x3, (lam * (x(P1) - x3) - y(P1)) % p)
+
+def point_mul(P: Optional[Point], n: int) -> Optional[Point]:
+    R = None
+    for i in range(256):
+        if (n >> i) & 1:
+            R = point_add(R, P)
+        P = point_add(P, P)
+    return R
+
+def bytes_from_int(x: int) -> bytes:
+    return x.to_bytes(32, byteorder="big")
+
+def bytes_from_point(P: Point) -> bytes:
+    return bytes_from_int(x(P))
+
+def lift_x(b: bytes) -> Optional[Point]:
+    x = int_from_bytes(b)
+    if x >= p:
+        return None
+    y_sq = (pow(x, 3, p) + 7) % p
+    y = pow(y_sq, (p + 1) // 4, p)
+    if pow(y, 2, p) != y_sq:
+        return None
+    return (x, y if y & 1 == 0 else p-y)
+
+def int_from_bytes(b: bytes) -> int:
+    return int.from_bytes(b, byteorder="big")
+
+def has_even_y(P: Point) -> bool:
+    assert not is_infinite(P)
+    return y(P) % 2 == 0
+
+#
+# End of helper functions copied from BIP-340 reference implementation.
+#
+
+def cbytes(P: Point) -> bytes:
+    a = b'\x02' if has_even_y(P) else b'\x03'
+    return a + bytes_from_point(P)
+
+def point_negate(P: Point) -> Point:
+    if is_infinite(P):
+        return P
+    return (x(P), p - y(P))
+
+def pointc(x: bytes) -> Point:
+    P = lift_x(x[1:33])
+    if x[0] == 2:
+        return P
+    elif x[0] == 3:
+        return point_negate(P)
+    assert False
+
+def key_agg(pubkeys: List[bytes]) -> bytes:
+    Q = key_agg_internal(pubkeys)
+    return bytes_from_point(Q)
+
+def key_agg_internal(pubkeys: List[bytes]) -> Point:
+    u = len(pubkeys)
+    Q = None
+    for i in range(u):
+        a_i = key_agg_coeff(pubkeys, pubkeys[i])
+        P_i = lift_x(pubkeys[i])
+        Q = point_add(Q, point_mul(P_i, a_i))
```

Comment:
@jonasnick This is roughly what I suggest for KeyAgg in the spec. It's the same as in our C code. I think we should give that algorithm to people, then they don't need to figure it out by themselves. It's not entirely trivial. 
```suggestion
    second = None
    for i in range(1, u):
        if pubkeys[i] != pubkeys[1]: 
            second = i
            break
    for i in range(u):
        a_i = 1 if i == second else key_agg_coeff(pubkeys, pubkeys[i])
        P_i = lift_x(pubkeys[i])
        Q = point_add(Q, point_mul(P_i, a_i))
```

@robot-dreams (not this PR, sorry for posting this here :P)

Reply:
Is your suggestion to remove the IsSecond predicate? I'm not able to spot the advantages of your approach immediately. I think it is attractive that `key_agg_coeff` is the coefficient of the key (including the second one). Another downside is that overwriting a variable (`second`) has been criticized in the past, which probably also applies to breaking.


Btw, it should be `pubkeys[i] != pubkeys[0]: `

Reply:
My reasoning is that this algorithm has running O(n) whereas the current one has O(n^2). Implementers will notice and try to come up with the more efficient version (our own C code is the best example!). I believe that one of the main reasons for having pseudocode and even Python code in cryptographic specs is that people can just translate it to their target language without the need to think about the algorithm (and potentially screw up, see my mistake... -.-).

> Is your suggestion to remove the IsSecond predicate? [...] I think it is attractive that `key_agg_coeff` is the coefficient of the key (including the second one). 

I agree that it's is attractive that `key_agg_coeff` is a pure function currently, it's certainly more readable (I think there's a reason why we wrote it like this in the paper.) Maybe we can have the best of both worlds: We can write IsSecond and  KeyAggCoeff in "mathematical" style such that it's easy to grasp and give algorithms for the more efficient version?

Reply:
Oh sorry, I hadn't considered performance optimization at all since we had only discussed so far how IsSecond is confusing people. I generally prefer clarity over performance, but in this case I agree that we don't want to have O(N^2) in the spec unnecessarily.

Keeping the IsSecond calculation inside KeyAggCoef is not only more readable but also required if we want signing and partial sig verification to not deal with IsSecond directly, because both algorithms call KeyAggCoef to get the coef of a single signer.

Providing both a naive algorithm and a performant one is an interesting idea to explore.

Reply:
By the way, I think the screw up potential when trying to optimize this yourself is neglible if we provide good test vectors.